parser-typechecker/src/Unison/Typechecker/Context.hs

```diff
@@ -867,75 +887,179 @@ noteTopLevelType e binding typ = case binding of
   _ -> btw $ topLevelComponent
     [(Var.reset (ABT.variable e), generalizeAndUnTypeVar typ, True)]
 
--- | Synthesize the type of the given term, updating the context in the process.
+synthesizeTop
+  :: Var v
+  => Ord loc
+  => Term v loc
+  -> M v loc (Type v loc)
+synthesizeTop tm = do
+  (ty, want) <- synthesize tm
+  ctx <- getContext
+  want <- substAndDefaultWanted want (out ctx)
+  when (not $ null want) . failWith $ do
+    AbilityCheckFailure
+      []
+      (Type.flattenEffects . snd =<< want)
+      ctx
+  applyM ty
+  where
+  out (Context es) = fmap fst es
+
+-- | Synthesize the type of the given term, updating the context in
+-- the process.  Also collect wanted abilities.
 -- | Figure 11 from the paper
-synthesize :: forall v loc . (Var v, Ord loc) => Term v loc -> M v loc (Type v loc)
-synthesize e | debugEnabled && traceShow ("synthesize"::String, e) False = undefined
+synthesize
+  :: Var v
+  => Ord loc
+  => Term v loc
+  -> M v loc (Type v loc, Wanted v loc)
+synthesize e | debugShow ("synthesize"::String, e) = undefined
 synthesize e = scope (InSynthesize e) $
   case minimize' e of
     Left es -> failWith (DuplicateDefinitions es)
     Right e -> do
-      Type.Effect'' es t <- go e
-      abilityCheck es
-      pure t
-  where
-  l = loc e
-  go :: (Var v, Ord loc) => Term v loc -> M v loc (Type v loc)
-  go (Term.Var' v) = getContext >>= \ctx -> case lookupAnn ctx v of -- Var
+      (Type.Effect'' es t, want) <- synthesizeWanted e
+      want <- coalesceWanted (fmap (Just e,) es) want
+      pure (t, want)
+
+wantRequest
+  :: Var v
+  => Ord loc
+  => Term v loc
+  -> Type v loc
+  -> (Type v loc, Wanted v loc)
+wantRequest loc ~(Type.Effect'' es t) = (t, fmap (Just loc,) es)
+
+synthesizeWanted
+  :: Var v
+  => Ord loc
+  => Term v loc
+  -> M v loc (Type v loc, Wanted v loc)
+synthesizeWanted (Term.Var' v) = getContext >>= \ctx ->
+  case lookupAnn ctx v of -- Var
     Nothing -> compilerCrash $ UndeclaredTermVariable v ctx
-    Just t -> pure t
-  go (Term.Blank' blank) = do
+    -- variables accesses are pure
+    Just t -> pure (t, [])
+synthesizeWanted (Term.Ref' h)
+  = compilerCrash $ UnannotatedReference h
+synthesizeWanted (Term.Ann' (Term.Ref' _) t)
+  -- innermost Ref annotation assumed to be correctly provided by
+  -- `synthesizeClosed`
+  --
+  -- Top level references don't have their own effects.
+  | Set.null s = (,[]) <$> existentializeArrows t
+  | otherwise = compilerCrash $ FreeVarsInTypeAnnotation s
+  where s = ABT.freeVars t
+synthesizeWanted (Term.Constructor' r cid)
+  -- Constructors do not have effects
+  = (,[]) . Type.purifyArrows <$> getDataConstructorType r cid
+synthesizeWanted tm@(Term.Request' r cid) =
+  fmap (wantRequest tm) . ungeneralize . Type.purifyArrows
+    =<< getEffectConstructorType r cid
```

Comment:
It took me a minute to figure out that this will still return `[]` if the request isn't fully saturated.